### PR TITLE
feat: promo-code-unlock patch

### PR DESCRIPTION
**Pull request by Technikte**
This adds a patch for the WarnWetter app, which enables the unlocking of all features via any string in promo code activation field.
I tested it and for it worked for me.

The patch and so on is a bit ugly since this is my first kotlin code and revanced patch. Thanks again to oSumAtrIX and canny.

Screenshot from App..
asdeasd

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,6 +20,14 @@ Official patches by ReVanced
 | `general-reddit-ads` | Removes general ads from the Reddit frontpage and subreddits. | all |
 </details>
 
+### 📦 `de.dwd.warnapp`
```
Comment (oSumAtrIX) [inline README.md:23]: Please do not commit this file, it is being generated by the workflow.

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/fingerprints/PromoCodeUnlockFingerprint.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import app.revanced.patcher.fingerprint.method.annotation.MatchingMethod
+import app.revanced.patches.youtube.layout.createbutton.annotations.CreateButtonCompatibility
+import org.jf.dexlib2.AccessFlags
+import org.jf.dexlib2.Opcode
+
+@Name("promo-code-unlock-fingerprint")
+@MatchingMethod(
+    "Lde/dwd/warnapp/model/PromoTokenVerification;", "isValid"
+)
+@FuzzyPatternScanMethod(2) // FIXME: Test this threshold and find the best value.
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/fingerprints/PromoCodeUnlockFingerprint.kt:17]:
```suggestion
import app.revanced.patcher.fingerprint.method.annotation.MatchingMethod
import app.revanced.patches.youtube.layout.createbutton.annotations.CreateButtonCompatibility
import org.jf.dexlib2.AccessFlags
import org.jf.dexlib2.Opcode

@Name("promo-code-unlock-fingerprint")
@MatchingMethod(
    "Lde/dwd/warnapp/model/PromoTokenVerification;", "isValid"
)
```

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patcher.patch.impl.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patches.warnwetter.misc.promocode.annotations.PromoCodeUnlockCompatibility
+import app.revanced.patches.warnwetter.misc.promocode.fingerprints.PromoCodeUnlockFingerprint
+
+@Patch
+@Name("promo-code-unlock")
+@Description("Disable the validation of promo code, any string will work to unlock all features.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:19]:
```suggestion
@Description("Disables the validation of promo code. Any code will work to unlock all features.")
```

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+@PromoCodeUnlockCompatibility
+@Version("0.0.1")
+class PromoCodeUnlockPatch : BytecodePatch(
+    listOf(
+        PromoCodeUnlockFingerprint
+    )
+) {
+
+    override fun execute(data: BytecodeData): PatchResult {
+        PromoCodeUnlockFingerprint.result!!.mutableMethod.removeInstructions(0, PromoCodeUnlockFingerprint.result!!.mutableMethod.implementation!!.instructions.size - 1)
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:29]: Please create variables.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt (hunk 3 above)
Comment (Technikte) [reply src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:19]: done

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/fingerprints/PromoCodeUnlockFingerprint.kt (hunk 2 above)
Comment (Technikte) [reply src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/fingerprints/PromoCodeUnlockFingerprint.kt:17]: done

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt (hunk 4 above)
Comment (Technikte) [reply src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:29]: I dont unstand.. Where and which var ?

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:29]: You have multiple occurrences of:
```kt
PromoCodeUnlockFingerprint.result!!
```
```kt
PromoCodeUnlockFingerprint.result!!.mutableMethod
```

Please create variables for them.

Comment (Technikte) [reply src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:29]: Alright, now I think understand what you mean..

File: src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+@Version("0.0.1")
+class PromoCodeUnlockPatch : BytecodePatch(
+    listOf(
+        PromoCodeUnlockFingerprint
+    )
+) {
+
+    override fun execute(data: BytecodeData): PatchResult {
+        val method = PromoCodeUnlockFingerprint.result!!.mutableMethod
+        method.removeInstructions(0, method.implementation!!.instructions.size - 1)
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/warnwetter/misc/promocode/patch/PromoCodeUnlockPatch.kt:30]: This is not required
```suggestion
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Comment by oSumAtrIX** [comment]
Thanks!

**Pull request merged**